### ISSUE TITLE
How to lose focus on clicked button?

### ISSUE THREAD

**Issue by rainyear** [react-bootstrap/react-bootstrap#1300]
Button stay focused after click:

![qq20150908-1 2x](https://cloud.githubusercontent.com/assets/3953966/9725382/471e04ce-5617-11e5-94b9-c6166b18f5e8.png)

the outline style can be removed by css: 

```css
.btn:focus {
  outline: none;
}
```

but the Button still get the focus when I press the Space key, and only Chrome has this issue, works fine on Safari.

**Comment by jquense** [comment]
this isn't an issue with react-bootstrap, it's the default browser behavior of Chrome.

**Comment by jquense** [comment]
I will also just leave this here for some context around removing outlines: http://outlinenone.com/

**Comment by munzx** [comment]
For anyone facing the same issue, just add this code anywhere on your page.
`<script>
	document.addEventListener('click', function(e) {
		document.activeElement.blur();
	});
</script>`

**Comment by jquense** [comment]
Please don't do that at all. forcing focus from buttons is actively hostile to users, especially those using the keyboard, SR, or any other non mouse input

**Comment by macgyver** [comment]
+1 for leaving focus where it is.

If you've got a better place to put the focus, e.g. within a modal dialog after clicking a button to open the modal, then go ahead and put it there, but don't blur a button just to clear the focus styles - it disrupts the experience of keyboard users.

**Comment by jfbloom22** [comment]
@jquense can you expand on how @munzx 's solution negatively affects those using the keyboard, SR, or other non mouse inputs?  In my tests with my keyboard, @munzx's solution is only removing the focus from a button when I click on it with my mouse, press the return key, or press the space bar while the button is focused.  I prefer this over simply adding CSS to remove the "focus" look entirely.   The only issue I can think of is if my application had a button that required the user to click multiple times in a row.   Then the user would have to re-focus the button each time.

**Comment by jfbloom22** [comment]
Found a solution that appears to prevent focusing on a button click and maintain good accessibility as well.  What do ya'll think?:

`onMouseDown={this.handleClick} 
 onKeyUp={(e) => {if (e.keyCode === 13 || e.keyCode === 32) {this.handleClick()}}}`

`handleClick(e) {
    if (e) {e.preventDefault()};
}`

The focusing of a button happens after the MouseDown event, so we prevent that when users click. If we stop there, then the space bar and return key do not "click" the button. Do not preventDefault for the onKeyUp event because then the button continues to look "focused" after clicking somewhere else.

**Comment by jquense** [comment]
the css/speced way to handle this is `:focus-visible` (see: https://github.com/WICG/focus-visible/blob/master/explainer.md) I'd suggest looking for polyfills for that if you don't like the current behavior 👍

**Comment by jfbloom22** [comment]
@jquense it would be awesome if the browsers implemented something similar to this.  One advantage I noticed my solution has, mine should still work for the NVDA screen reader.  https://github.com/WICG/focus-visible/issues/128

I am sure my solution has issues, but I do not have any assistive technology to test with.  At this point all I know is that it works well for the Keyboard and should not have that one specific issue I linked to above.

**Comment by Hsia1993** [comment]
###

**Comment by fuzzbomb** [comment]
WHY do you want to prevent focus when a mouse user clicks? No-one has mentioned a use case here.

@jquense is correct. This is hostile to users.

Programatically preventing a button from gaining focus when clicked, likely qualifies as a failure of WCAG success criterion [2.5.6 Concurrent Input Mechanisms](https://www.w3.org/WAI/WCAG21/Understanding/concurrent-input-mechanisms). _"Users should be able to switch input mechanisms at any point should the user determine that certain tasks and interactions are more easily accomplished by using an alternative input mechanism."_

Pointer users DO benefit from focus:
1. When clicking a button (or link, or other control), a visible focus indicator offers _confirmation that you hit the the control you intended_. This is can be helpful in situations where:
    - There are multiple controls close together. For example a group of toolbar buttons, or a set of checkboxes.
    - Using a coarse pointer.
    - You aren't very accurate or confident with the pointer. For example, because this isn't your own device, or it isn't equipped with your preferred kind of pointer.
    - Users have a variety of cognitive impairments. For example distractability or working memory impairment.
2. When using mouse and keyboard in combination:
    - You can click one control, then continue by pressing TAB to move to subsequent controls. For example, you can use a pointer click to open a collapsible group (details/summary), then tab through the controls they find inside. *Programatically preventing the control from gaining focus after a mouse click will confound this.*
    - Some people who prefer keyboard tabbing may occasionally resort to a pointer, even if using the pointer isn't comfortable. For example as a shortcut to avoid a long number of tab presses, or when the keyboard tabbing sequence is hard to follow.

It's a mistake to treat pointer users and keyboard users as entirely separate groups of people. In reality, many people make use of both the mouse pointer and keyboard tabbing.

**Comment by ephracis** [comment]
One use case would be when you have an editable element with a selection, and you have a button to toggle formatting, for example bold. Then the clicking of the "Toggle Bold" button steals focus and removes the selected text inside the editable element, so the function `document.execCommand` does nothing.

This is what I'm struggling with, but maybe I am missing something here?

**Comment by mschwartz** [comment]
My use case is an application that runs 24/7/365 on a tablet mounted on the wall.  There is no keyboard, no need for any sort of browser assist.  Ideally, the UI would show button press and unpress, without remaining highlighted in any way.

I don't suggest that removing focus is active state from elements is generally a good idea, but there's always an exception.

**Comment by kElias-lrw** [comment]
My use case is taking a survey, the button should remove focus for the next question so as not to bias respondents.

**Comment by mschwartz** [comment]
Not fixed since 2015?  Please :)

**Comment by Drekdeus** [comment]
My solution [React - TypeScript]:
` <button onFocus={(e:any) => (e.target.blur())}`

**Comment by eli7gn** [comment]
> onFocus={(e:any) => (e.target.blur())}

Thanks a lot!!!!  I did: onFocus={(e) => e.target.blur()}

**Comment by fpdude** [comment]
> My solution [React - TypeScript]: ` <button onFocus={(e:any) => (e.target.blur())}`

Thanks, that's a good solution for my case!

**Comment by jquense** [comment]
folks once again, DO NOT blur buttons on focus. It is a huge anti-pattern and use hostile. If you are very unhappy with this override the focus styles of button to work on `focus-visible` there is even a good focus-visible polyfill for old browser. 

PLEASE, FOR THE SAKE OF YOUR USERS, DO THIS THE CORRECT WAY, 

Thank you
